### PR TITLE
Initial ASR to Fortran implementation

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Towards: https://github.com/lfortran/lfortran/issues/2410

### CODE REVIEW

**Comment by Pranavchiku** [comment]
> Save it into a file and compile it using GFortran in `integration_tests/run_tests`?


I feel this can be a better option as we will be sure that generated code is correct. Although I do not have entire context of what, why and how things need to be done.

We may even test with LFortran as well, generate an ASR and see if it matches node by node ( except symtab IDs ).

**Comment by certik** [comment]
I would test it exactly as we test the C backend. So both tests in `tests` and integration tests, it would be `--backend=gfortran` most likely.

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+
+    /********************************** Symbol *********************************/
+    void visit_Program(const ASR::Program_t &x) {
+        std::string r;
+        r = "program";
+        r += " ";
+        r.append(x.m_name);
+        r += "\n";
+        inc_indent();
+        r += indent + "implicit none"; // TODO: Handle implicit
```
Comment (certik) [inline src/libasr/codegen/asr_to_fortran.cpp:89]: There is no implicit typing in ASR, so you just put in `implicit none` and be done with it.

**Comment by certik** [comment]
Looks good so far!

**Comment by Thirumalai-Shaktivel** [comment]
I'm getting `/home/runner/micromamba/envs/gf/bin/../lib/gcc/x86_64-conda-linux-gnu/13.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/runner/work/lfortran/lfortran/src/bin/../runtime/liblfortran_runtime.so: undefined reference to clock_gettime@GLIBC_2.17` in Linux. After fixing it, this PR is ready for review!

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -806,6 +806,51 @@ jobs:
             ./run_tests.py -b wasm
             ./run_tests.py -b wasm -f
 
+  test_gfortran:
```
Comment (certik) [inline .github/workflows/CI.yml:809]:
```suggestion
  test_fortran:
```

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -806,6 +806,51 @@ jobs:
...
+    name: Test GFortran backend
```
Comment (certik) [inline .github/workflows/CI.yml:810]:
```suggestion
    name: Test Fortran backend
```

File: src/bin/lfortran.cpp [hunk 4]
```diff
@@ -64,7 +64,7 @@ using LCompilers::endswith;
 using LCompilers::CompilerOptions;
 
 enum Backend {
-    llvm, c, cpp, x86, wasm
+    llvm, c, cpp, x86, wasm, gfortran
```
Comment (certik) [inline src/bin/lfortran.cpp:67]:
```suggestion
    llvm, c, cpp, x86, wasm, fortran
```

File: src/bin/lfortran.cpp [hunk 5]
```diff
@@ -1403,6 +1426,43 @@ int compile_to_object_file_c(const std::string &infile,
     return 0;
 }
 
+int compile_to_binary_gfortran(const std::string &infile,
```
Comment (certik) [inline src/bin/lfortran.cpp:1429]:
```suggestion
int compile_to_binary_fortran(const std::string &infile,
```

File: src/bin/lfortran.cpp [hunk 6]
```diff
@@ -1403,6 +1426,43 @@ int compile_to_object_file_c(const std::string &infile,
...
+    }
+
+    std::string in_file = outfile + ".tmp.f90";
+    {
+        std::ofstream out;
+        out.open(in_file);
+        out << src.result;
+    }
+
+    std::string cmd = "gfortran -o " + outfile + " -c " + in_file;
```
Comment (certik) [inline src/bin/lfortran.cpp:1457]: Later we'll make it configurable. We should probably use `lfortran` as the default. But for now gfortran is fine.

File: src/bin/lfortran.cpp [hunk 7]
```diff
@@ -1602,6 +1662,17 @@ int link_executable(const std::vector<std::string> &infiles,
             return 10;
         }
         return 0;
+    } else if (backend == Backend::gfortran) {
```
Comment (certik) [inline src/bin/lfortran.cpp:1665]:
```suggestion
    } else if (backend == Backend::fortran) {
```

File: src/bin/lfortran.cpp [hunk 8]
```diff
@@ -1900,14 +1972,15 @@ int main(int argc, char *argv[])
...
+        app.add_flag("--show-fortran", show_fortran, "Show Fortran translation source for the given file and exit");
         app.add_flag("--show-stacktrace", compiler_options.show_stacktrace, "Show internal stacktrace on compiler errors");
         app.add_flag("--symtab-only", compiler_options.symtab_only, "Only create symbol tables in ASR (skip executable stmt)");
         app.add_flag("--time-report", time_report, "Show compilation time report");
         app.add_flag("--static", static_link, "Create a static executable");
         app.add_flag("--no-warnings", compiler_options.no_warnings, "Turn off all warnings");
         app.add_flag("--no-error-banner", compiler_options.no_error_banner, "Turn off error banner");
         app.add_option("--error-format", compiler_options.error_format, "Control how errors are produced (human, short)")->capture_default_str();
-        app.add_option("--backend", arg_backend, "Select a backend (llvm, cpp, x86, wasm)")->capture_default_str();
+        app.add_option("--backend", arg_backend, "Select a backend (llvm, cpp, x86, wasm, gfortran)")->capture_default_str();
```
Comment (certik) [inline src/bin/lfortran.cpp:1983]:
```suggestion
        app.add_option("--backend", arg_backend, "Select a backend (llvm, cpp, x86, wasm, fortran)")->capture_default_str();
```

File: src/bin/lfortran.cpp [hunk 9]
```diff
@@ -2049,8 +2122,10 @@ int main(int argc, char *argv[])
             backend = Backend::x86;
         } else if (arg_backend == "wasm") {
             backend = Backend::wasm;
+        } else if (arg_backend == "gfortran") {
```
Comment (certik) [inline src/bin/lfortran.cpp:2125]:
```suggestion
        } else if (arg_backend == "fortran") {
```

File: tests/reference/fortran-expr5-3a279bb.stdout [hunk 10]
```diff
@@ -0,0 +1,9 @@
+program expr_05
+    implicit none
+    integer(4) :: x
+
+    x = 2 + 3 * 5
```
Comment (certik) [inline tests/reference/fortran-expr5-3a279bb.stdout:5]: The parentheses are incorrect.

**Review by certik — COMMENTED** [review]
I think this looks good in general. The main comments I left have to do with the fact that we should think of this as the Fortran backend, just like we think of the C backend as a C backend, not a GCC or Clang backend.

Furthermore, we should later on test this with LFortran itself, in addition to GFortran. The generated code should work with any Fortran compiler.

Also, we'll have to later provide options for various dialects. For example, the generics will only work with LFortran. I think the easiest will be an option "standard fortran", which will make this backend to instantiate templates and perhaps provide the generic code in comments, and it will use such subset of Fortran that compiles in all Fortran compilers (GFortran, LFortran, Intel IFort, NAG, etc.). This will be useful to allow using LFortran extensions easily with other compilers.

**Review by czgdp1807 — APPROVED** [review]
LGTM. Some refactorings can happen in the backend like a common function for generating strings for types, intents, and macros for visiting symbol tables. But that should be done in follow up PRs and is not very high priority.

**Review by certik — APPROVED** [review]
I think this looks great and good enough to merge as the initial implementation. Thanks @Thirumalai-Shaktivel !

I would initially do it without color, as it should be easier and maybe we don't need color here at all. The indenting we'll make configurable like we did in AST->SRC.

**Comment by Thirumalai-Shaktivel** [comment]
Yup, let's do it.
Thanks for reviewing and merging the PR.

**Pull request merged**